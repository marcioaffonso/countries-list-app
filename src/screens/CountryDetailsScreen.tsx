import React, { useEffect } from 'react';
import { View, Text, StyleSheet, RefreshControl, ScrollView } from 'react-native';
import { observer } from 'mobx-react-lite';
import PropTypes from 'prop-types';
import countryStore from '../stores/countryStore';
import CustomLoader from '../components/CustomLoader';

const CountryDetailsScreen = ({ route }) => {
  const { countryName } = route.params;
  const { loadingDetails, refreshingDetails, countryDetails } = countryStore;
  const loadCountryInfo = () => {
    countryStore.loadCountryDetails(countryName);
  }
  useEffect(loadCountryInfo, [countryName]);
  return (
    <ScrollView
      contentContainerStyle={styles.container}
      refreshControl={
        <RefreshControl refreshing={refreshingDetails} onRefresh={loadCountryInfo} />
      }
    >
      { loadingDetails || !countryDetails ? <CustomLoader /> :
        <>
          <Text style={styles.flag}>{ countryDetails.flag }</Text>
          <Text style={styles.name}>{ countryDetails.name.common }</Text>
          <View style={styles.details}>
            <Text style={styles.detailsText}>üåé Name: { countryDetails.name.official }</Text>
            <Text style={styles.detailsText}>üè† Capital: { countryDetails.capital.join(', ') }</Text>
            <Text style={styles.detailsText}>
              üíµ Currency: { Object.keys(countryDetails.currencies).map((key) => {
                const currency = countryDetails.currencies[key];
                return `${key} (${currency.symbol}) - ${currency.name}`;
              }).join(', ') }
            </Text>
          </View>
        </>
      }
    </ScrollView>
  );
};

CountryDetailsScreen.propTypes = {
  route: PropTypes.shape({
    params: PropTypes.any,
  }).isRequired,
};

export default observer(CountryDetailsScreen);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#eee',
    alignItems: 'center',
  },
  name: {
    fontSize: 40,
    textAlign: 'center',
  },
  flag: {
    textAlign: 'center',
    fontSize: 200
  },
  details: {
    marginTop: 60,
    paddingHorizontal: 10
  },
  detailsText: {
    fontSize: 24
  }
});
