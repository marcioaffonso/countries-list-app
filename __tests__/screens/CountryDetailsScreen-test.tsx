import React from 'react';
import renderer from 'react-test-renderer';
import { jest, expect, test } from '@jest/globals';
import CountryDetailsScreen from '../../src/screens/CountryDetailsScreen';
import { mockCountries }  from '../mocks/mockCountries';

const mockRouteParams = {
  params: {
    country: mockCountries[0].name.common
  }
};

jest.mock('../../src/stores/countryStore', () => {
  return {
    loadingDetails: false,
    refreshingDetails: false,
    countryDetails: {
      name: { common: 'Country1', official: 'Official1' },
      flag: 'ðŸ‡§ðŸ‡·',
      capital: ['Capital1'],
      currencies: { "C1": { name: "currenycy1", symbol: '$' }}
    },
    loadCountryDetails: jest.fn()
  };
});

test('renders country details correctly', () => {
  const tree = renderer.create(<CountryDetailsScreen route={mockRouteParams} />).toJSON();
  expect(tree).toMatchSnapshot();
});
