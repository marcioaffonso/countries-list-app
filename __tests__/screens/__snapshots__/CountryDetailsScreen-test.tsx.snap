// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders country details correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "alignItems": "center",
      "backgroundColor": "#eee",
      "flex": 1,
    }
  }
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
>
  <RCTRefreshControl />
  <View>
    <Text
      style={
        {
          "fontSize": 200,
          "textAlign": "center",
        }
      }
    >
      🇧🇷
    </Text>
    <Text
      style={
        {
          "fontSize": 40,
          "textAlign": "center",
        }
      }
    >
      Country1
    </Text>
    <View
      style={
        {
          "marginTop": 60,
          "paddingHorizontal": 10,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 24,
          }
        }
      >
        🌎 Name: 
        Official1
      </Text>
      <Text
        style={
          {
            "fontSize": 24,
          }
        }
      >
        🏠 Capital: 
        Capital1
      </Text>
      <Text
        style={
          {
            "fontSize": 24,
          }
        }
      >
        💵 Currency: 
        C1 ($) - currenycy1
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
